relayer/chains/wasm/provider.go

```diff
@@ -570,41 +573,147 @@ func (p *Provider) fetchBlockMessages(ctx context.Context, heightInfo *types.Hei
+		go func(event sdkTypes.Event) {
 			defer wg.Done()
-			searchParam.Events = append(searchParam.Events, event)
-			res, err := p.client.TxSearch(ctx, searchParam)
-			if err != nil {
-				errorChan <- err
-				return
+
+			var (
+				localSearchParam = searchParam
+				localMessages    = new(coreTypes.ResultTxSearch)
+				retryCount       uint8
+				zapFields        = []zap.Field{
+					zap.Uint64("start_height", localSearchParam.StartHeight),
+					zap.Uint64("end_height", localSearchParam.EndHeight),
+					zap.String("event", event.Type),
+				}
+			)
+			localSearchParam.Events = append(localSearchParam.Events, event)
+
+			for retryCount < types.RPCMaxRetryAttempts {
+				select {
+				case <-ctx.Done():
+					return
+				default:
+				}
+
+				p.logger.Info("fetching block messages", zapFields...)
+				msgs, err := p.client.TxSearch(ctx, localSearchParam)
+				if err == nil {
+					p.logger.Info("fetched block messages", zapFields...)
+					localMessages = msgs
+					break
+				}
+				retryCount++
+				if retryCount >= types.RPCMaxRetryAttempts {
+					p.logger.Error("fetchBlockMessages failed", append(zapFields, zap.Uint8("attempt", retryCount), zap.Error(err))...)
+					select {
+					case errorChan <- err:
+					default:
+					}
+					cancel()
+					return
+				}
+				delay := time.Duration(math.Pow(2, float64(retryCount))) * types.BaseRPCRetryDelay
+				if delay > types.MaxRPCRetryDelay {
+					delay = types.MaxRPCRetryDelay
+				}
+				p.logger.Warn("failed to fetch block messages. retrying...", append(zapFields, zap.Uint8("attempt", retryCount), zap.Duration("retrying_in", delay), zap.Error(err))...)
+				select {
+				case <-ctx.Done():
+					return
+				case <-time.After(delay):
+				}
 			}
-			if res.TotalCount > perPage {
-				for i := 2; i <= int(res.TotalCount/perPage)+1; i++ {
-					searchParam.Page = &i
-					resNext, err := p.client.TxSearch(ctx, searchParam)
-					if err != nil {
-						errorChan <- err
+
+			if localMessages.TotalCount > perPage {
+				totalPages := (localMessages.TotalCount + perPage - 1) / perPage
+				for i := 2; i <= totalPages; i++ {
+					select {
+					case <-ctx.Done():
 						return
+					default:
+					}
+					p.logger.Info("fetching block messages", append(zapFields, zap.Int("page", i))...)
+					localSearchParam.Page = &i
+					resNext, err := p.client.TxSearch(ctx, localSearchParam)
+					if err != nil {
+						var attempts uint8
+						for attempts < types.RPCMaxRetryAttempts {
+							select {
+							case <-ctx.Done():
+								return
+							default:
+							}
+							p.logger.Warn("failed to fetch block messages with page", append(zapFields, zap.Int("page", i), zap.Uint8("attempt", attempts), zap.Error(err))...)
+							attempts++
+							if attempts >= types.RPCMaxRetryAttempts {
+								p.logger.Error("fetchBlockMessages failed", append(zapFields, zap.Int("page", i), zap.Uint8("attempt", attempts), zap.Error(err))...)
+								select {
```

Comment:
Can you refactor the implementation a bit so it would be more readable ?

Reply:
I'm not sure I understand the refactor you want.

more context or explanation?

Reply:
This was for the whole function. The function has grown significantly and it might be better to break down

Reply:
I understand that's why I have added TODO marker on function doc.

Reply:
Can't we refactor this now instead of deferring for later ??

Reply:
There's significant refactoring needed for overall relayer codebase.

That's the perfect moment.

Reply:
At the very least we can remove unwanted for { select {}} loops or unwanted empty cases.
```

select {
		case errorChan <- err:
		default:
	}


select {
		case <-ctx.Done():
			return
		case <-time.After(delay):
	}

```

Reply:
That is needed for cancellation.